float64 initial_error
float64 final_error
int64 size